lib/view_component/compiler.rb

```diff
@@ -38,46 +34,39 @@ def compile(raise_errors: false, force: false)
       end
 
       component_class.superclass.compile(raise_errors: raise_errors) if should_compile_superclass?
+      subclass_instance_methods = component_class.instance_methods(false)
 
-      with_write_lock do
-        CompileCache.invalidate_class!(component_class)
-
-        subclass_instance_methods = component_class.instance_methods(false)
-
-        if subclass_instance_methods.include?(:with_content) && raise_errors
-          raise ViewComponent::ComponentError.new(
-            "#{component_class} implements a reserved method, `#with_content`.\n\n" \
-            "To fix this issue, change the name of the method."
-          )
-        end
-
-        if template_errors.present?
-          raise ViewComponent::TemplateError.new(template_errors) if raise_errors
+      if subclass_instance_methods.include?(:with_content) && raise_errors
+        raise ViewComponent::ComponentError.new(
+          "#{component_class} implements a reserved method, `#with_content`.\n\n" \
+          "To fix this issue, change the name of the method."
+        )
+      end
 
-          return false
-        end
+      if template_errors.present?
+        raise ViewComponent::TemplateError.new(template_errors) if raise_errors
 
-        if subclass_instance_methods.include?(:before_render_check)
-          ViewComponent::Deprecation.warn(
-            "`#before_render_check` will be removed in v3.0.0.\n\n" \
-            "To fix this issue, use `#before_render` instead."
-          )
-        end
+        return false
+      end
 
-        if raise_errors
-          component_class.validate_initialization_parameters!
-          component_class.validate_collection_parameter!
-        end
+      if subclass_instance_methods.include?(:before_render_check)
+        ViewComponent::Deprecation.warn(
+          "`#before_render_check` will be removed in v3.0.0.\n\n" \
+          "To fix this issue, use `#before_render` instead."
+        )
+      end
 
-        templates.each do |template|
-          # Remove existing compiled template methods,
-          # as Ruby warns when redefining a method.
-          method_name = call_method_name(template[:variant])
+      if raise_errors
+        component_class.validate_initialization_parameters!
+        component_class.validate_collection_parameter!
+      end
 
-          if component_class.instance_methods.include?(method_name.to_sym)
-            component_class.send(:undef_method, method_name.to_sym)
-          end
+      templates.each do |template|
+        # Remove existing compiled template methods,
+        # as Ruby warns when redefining a method.
+        method_name = call_method_name(template[:variant])
 
+        silence_redefinition_of_methods(method_name) do
```

Comment:
I know we generally want less locks, but I think this could solve our multi-threading issue with `silence_redefinition_of_method`:

```ruby
# somewhere above we do: @redefinition_lock = Thread::Mutex.new
@redefinition_lock.synchronize do
  silence_redefinition_of_method(method_name)
end
```

This should be safe since we don't hold the lock in a (potentially) recursively called method like we did with templates, and there's no state like `CompileCache` that can cause deadlocks.

Thoughts on going with this approach to avoid having to track/call `silence_redefinition_of_method` per-thread?

Reply:
I was skeptical at first (because the whole point of this PR is to _remove_ locks 😅 ) but it appears to work 👍 It makes sense to me why it works too, since we're still not undefining any methods before defining them, thereby avoiding the race condition.